File: gitdot-server/src/handler/repository.rs
mod convert_readonly_repository;
mod create_repository;
mod create_repository_commit_filter;
mod delete_repository;
mod delete_repository_commit_filter;
mod get_repository;
mod get_repository_activity;
mod get_repository_blob;
mod get_repository_blob_diffs;
mod get_repository_blobs;
mod get_repository_commit;
mod get_repository_commit_diff;
mod get_repository_paths;
mod get_repository_resources;
mod list_latest_repositories;
mod list_repository_commit_filters;
mod list_repository_commits;
mod list_trending_repositories;
mod star_repository;
mod unstar_repository;
mod update_repository_commit_filter;

use axum::{
    Router,
    routing::{get, patch, post},
};

use crate::app::AppState;

use convert_readonly_repository::convert_readonly_repository;
use create_repository::create_repository;
use create_repository_commit_filter::create_repository_commit_filter;
use delete_repository::delete_repository;
use delete_repository_commit_filter::delete_repository_commit_filter;
use get_repository::get_repository;
use get_repository_activity::get_repository_activity;
use get_repository_blob::get_repository_blob;
use get_repository_blob_diffs::get_repository_blob_diffs;
use get_repository_blobs::get_repository_blobs;
use get_repository_commit::get_repository_commit;
use get_repository_commit_diff::get_repository_commit_diff;
use get_repository_paths::get_repository_paths;
use get_repository_resources::get_repository_resources;
use list_latest_repositories::list_latest_repositories;
use list_repository_commit_filters::list_repository_commit_filters;
use list_repository_commits::list_repository_commits;
use list_trending_repositories::list_trending_repositories;
use star_repository::star_repository;
use unstar_repository::unstar_repository;
use update_repository_commit_filter::update_repository_commit_filter;

pub fn create_repository_router() -> Router<AppState> {
    Router::new()
        .route("/repository/latest", get(list_latest_repositories))
        .route("/repository/trending", get(list_trending_repositories))
        .route(
            "/repository/{owner}/{repo}",
            post(create_repository)
                .get(get_repository)
                .delete(delete_repository),
        )
        .route("/repository/{owner}/{repo}/blob", get(get_repository_blob))
        .route(
            "/repository/{owner}/{repo}/blob/diffs",
            post(get_repository_blob_diffs),
        )
        .route(
            "/repository/{owner}/{repo}/blobs",
            post(get_repository_blobs),
        )
        .route(
            "/repository/{owner}/{repo}/paths",
            get(get_repository_paths),
        )
        .route(
            "/repository/{owner}/{repo}/commits",
            get(list_repository_commits),
        )
        .route(
            "/repository/{owner}/{repo}/commits/{sha}",
            get(get_repository_commit),
        )
        .route(
            "/repository/{owner}/{repo}/commits/{sha}/diff",
            get(get_repository_commit_diff),
        )
        .route(
            "/repository/{owner}/{repo}/resources",
            post(get_repository_resources),
        )
        .route("/repository/{owner}/{repo}/star", post(star_repository))
        .route("/repository/{owner}/{repo}/unstar", post(unstar_repository))
        .route(
            "/repository/{owner}/{repo}/activity",
            get(get_repository_activity),
        )
        .route(
            "/repository/{owner}/{repo}/convert-readonly",
            post(convert_readonly_repository),
        )
        .route(
            "/repository/{owner}/{repo}/commit-filters",
            post(create_repository_commit_filter).get(list_repository_commit_filters),
        )
        .route(
            "/repository/{owner}/{repo}/commit-filters/{filter_id}",
            patch(update_repository_commit_filter).delete(delete_repository_commit_filter),
        )
}

File: gitdot-web/packages/gitdot-client/src/repository.ts
import "server-only";

import {
  ConvertReadonlyRepositoryResponse,
  type CreateRepositoryCommitFilterRequest,
  CreateRepositoryCommitFilterResponse,
  type CreateRepositoryRequest,
  GetRepositoryActivityResponse,
  type GetRepositoryBlobDiffsRequest,
  type GetRepositoryBlobRequest,
  type GetRepositoryPathsRequest,
  type GetRepositoryResourcesRequest,
  ListLatestRepositoriesResponse,
  ListRepositoryCommitFiltersResponse,
  type ListRepositoryCommitsRequest,
  ListRepositoryCommitsResponse,
  ListTrendingRepositoriesResponse,
  RepositoryBlobDiffsResource,
  RepositoryBlobResource,
  RepositoryCommitDiffResource,
  type RepositoryCommitFilterResource,
  RepositoryCommitResource,
  RepositoryPathsResource,
  RepositoryResource,
  RepositoryResourcesResource,
  type UpdateRepositoryCommitFilterRequest,
  UpdateRepositoryCommitFilterResponse,
  type UpdateRepositoryRequest,
  UpdateRepositoryResponse,
} from "gitdot-api";
import {
  authDelete,
  authFetch,
  authPatch,
  authPost,
  GITDOT_SERVER_URL,
  handleEmptyResponse,
  handleResponse,
  toQueryString,
} from "./util";

export async function createRepository(
  owner: string,
  repo: string,
  request: CreateRepositoryRequest,
): Promise<RepositoryResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
    request,
  );

  return await handleResponse(response, RepositoryResource);
}

export async function updateRepository(
  owner: string,
  repo: string,
  request: UpdateRepositoryRequest,
): Promise<RepositoryResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
    request,
  );

  return await handleResponse(response, UpdateRepositoryResponse);
}

export async function getRepositoryBlob(
  owner: string,
  repo: string,
  query: GetRepositoryBlobRequest,
): Promise<RepositoryBlobResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/blob?${queryString}`,
  );

  return await handleResponse(response, RepositoryBlobResource);
}

// TODO: support pagination for repositories that need more than 10K
export async function listRepositoryCommits(
  owner: string,
  repo: string,
  query?: ListRepositoryCommitsRequest,
): Promise<ListRepositoryCommitsResponse | null> {
  const queryString = toQueryString(query);
  const url = `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commits${queryString ? `?${queryString}` : ""}`;
  const response = await authFetch(url);
  return await handleResponse(response, ListRepositoryCommitsResponse);
}

export async function listLatestRepositories(): Promise<ListLatestRepositoriesResponse | null> {
  const response = await authFetch(`${GITDOT_SERVER_URL}/repository/latest`);
  return await handleResponse(response, ListLatestRepositoriesResponse);
}

export async function listTrendingRepositories(): Promise<ListTrendingRepositoriesResponse | null> {
  const response = await authFetch(`${GITDOT_SERVER_URL}/repository/trending`);
  return await handleResponse(response, ListTrendingRepositoriesResponse);
}

export async function getRepositoryCommit(
  owner: string,
  repo: string,
  sha: string,
): Promise<RepositoryCommitResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commits/${sha}`,
  );

  return await handleResponse(response, RepositoryCommitResource);
}

export async function getRepositoryPaths(
  owner: string,
  repo: string,
  query?: GetRepositoryPathsRequest,
): Promise<RepositoryPathsResource | null> {
  const queryString = toQueryString(query);
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/paths?${queryString}`,
  );
  return await handleResponse(response, RepositoryPathsResource);
}

export async function listRepositoryCommitFilters(
  owner: string,
  repo: string,
  opts?: { cursor?: string; limit?: number },
): Promise<ListRepositoryCommitFiltersResponse | null> {
  const qs = toQueryString({ cursor: opts?.cursor, limit: opts?.limit });
  const url = `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commit-filters${qs ? `?${qs}` : ""}`;
  const response = await authFetch(url);
  return await handleResponse(response, ListRepositoryCommitFiltersResponse);
}

export async function createRepositoryCommitFilter(
  owner: string,
  repo: string,
  request: CreateRepositoryCommitFilterRequest,
): Promise<RepositoryCommitFilterResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commit-filters`,
    request,
  );
  return await handleResponse(response, CreateRepositoryCommitFilterResponse);
}

export async function updateRepositoryCommitFilter(
  owner: string,
  repo: string,
  filterId: string,
  request: UpdateRepositoryCommitFilterRequest,
): Promise<RepositoryCommitFilterResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commit-filters/${filterId}`,
    request,
  );
  return await handleResponse(response, UpdateRepositoryCommitFilterResponse);
}

export async function deleteRepositoryCommitFilter(
  owner: string,
  repo: string,
  filterId: string,
): Promise<void> {
  const response = await authDelete(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commit-filters/${filterId}`,
  );
  await handleEmptyResponse(response);
}

export async function getRepositoryBlobDiffs(
  owner: string,
  repo: string,
  request: GetRepositoryBlobDiffsRequest,
): Promise<RepositoryBlobDiffsResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/blob/diffs`,
    request,
  );
  return await handleResponse(response, RepositoryBlobDiffsResource);
}

export async function getRepositoryCommitDiff(
  owner: string,
  repo: string,
  sha: string,
): Promise<RepositoryCommitDiffResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/commits/${sha}/diff`,
  );
  return await handleResponse(response, RepositoryCommitDiffResource);
}

export async function getRepository(
  owner: string,
  repo: string,
): Promise<RepositoryResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
  );
  return await handleResponse(response, RepositoryResource);
}

export async function getRepositoryActivity(
  owner: string,
  repo: string,
): Promise<GetRepositoryActivityResponse | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/activity`,
  );
  return await handleResponse(response, GetRepositoryActivityResponse);
}

export async function getRepositoryResources(
  owner: string,
  repo: string,
  request: GetRepositoryResourcesRequest = {},
): Promise<RepositoryResourcesResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/resources`,
    request,
  );
  return await handleResponse(response, RepositoryResourcesResource);
}

export async function deleteRepository(
  owner: string,
  repo: string,
): Promise<void> {
  const response = await authDelete(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}`,
  );
  if (!response.ok) {
    let message = response.statusText;
    try {
      const body = await response.json();
      if (typeof body?.message === "string") message = body.message;
    } catch {}
    throw new Error(message);
  }
}

export async function starRepository(
  owner: string,
  repo: string,
): Promise<void> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/star`,
    {},
  );
  await handleEmptyResponse(response);
}

export async function unstarRepository(
  owner: string,
  repo: string,
): Promise<void> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/unstar`,
    {},
  );
  await handleEmptyResponse(response);
}

export async function convertReadonlyRepository(
  owner: string,
  repo: string,
): Promise<ConvertReadonlyRepositoryResponse | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/convert-readonly`,
    {},
  );
  return await handleResponse(response, ConvertReadonlyRepositoryResponse);
}
